src/codec.rs

```diff
@@ -885,9 +933,118 @@ impl<T: Encode, const N: usize> Encode for [T; N] {
 }
 
 impl<T: Decode, const N: usize> Decode for [T; N] {
-	#[inline]
+	#[inline(always)]
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		decode_array(input)
+		let mut array = MaybeUninit::uninit();
+		Self::decode_into(DecodeContext::new(), input, &mut array)?;
+
+		// SAFETY: `decode_into` succeeded, so the array is initialized.
+		unsafe {
+			Ok(array.assume_init())
+		}
+	}
+
+    fn decode_into<I: Input>(ctx: DecodeContext, input: &mut I, dst: &mut MaybeUninit<Self>) -> Result<DecodeFinished, Error> {
+		let is_primitive = match <T as Decode>::TYPE_INFO {
+			| TypeInfo::U8
+			| TypeInfo::I8
+				=> true,
+			| TypeInfo::U16
+			| TypeInfo::I16
+			| TypeInfo::U32
+			| TypeInfo::I32
+			| TypeInfo::U64
+			| TypeInfo::I64
+			| TypeInfo::U128
+			| TypeInfo::I128
+			| TypeInfo::F32
+			| TypeInfo::F64
+				=> cfg!(target_endian = "little"),
+			TypeInfo::Unknown => false
+		};
+
+		if is_primitive {
+			// Let's read the array in bulk as that's going to be a lot
+			// faster than just reading each element one-by-one.
+
+			let ptr: *mut [T; N] = dst.as_mut_ptr();
+			let ptr: *mut u8 = ptr.cast();
+			let bytesize = core::mem::size_of::<T>().checked_mul(N).expect("array lengths are sane");
+
+			// TODO: This is potentially slow; it'd be better if `Input` supported
```

Comment:
Hm, sorry, I'm not sure I understand what you're suggesting here. What exactly would this change? (: They'd have to be zeroed anyway as long as `Input` accepts a `&[u8]` instead of `&mut [MaybeUninit<u8>]`.

Reply:
> They'd have to be zeroed anyway

Yes, but they will be zeroed by compiler and there will be no need for an unsafe code to memzero them.

The previous implementation used that trick for i8/u8, see `let mut array: [u8; N] = [0; N];`. Still this comment is not a blocker, you can ignore it/implement in another PR.

Reply:
Okay, I see what you mean now!

Hmm... it which case it'd probably be better to just make it possible to read into the array without initializing it. (: